Library/Homebrew/cmd/update.sh

```diff
@@ -0,0 +1,242 @@
+    then
+      echo "Your HEAD is not a descendant of $UPSTREAM_BRANCH!" >&2
+      exit 1
+    fi
+    return
+  fi
+
+  if [ "$INITIAL_BRANCH" != "$UPSTREAM_BRANCH" ] && [ -n "$INITIAL_BRANCH" ]
+  then
+    git checkout --force -B "$UPSTREAM_BRANCH" "origin/$UPSTREAM_BRANCH" $QUIET_ARGS
+  fi
+
+  INITIAL_REVISION="$(read_current_revision)"
+
+  # ensure we don't munge line endings on checkout
+  git config core.autocrlf false
+
+  trap reset_on_interrupt SIGINT
+
+  if [ -n "$HOMEBREW_REBASE" ]
+  then
+    git rebase $QUIET_ARGS "origin/$UPSTREAM_BRANCH"
+  else
+    git merge --no-edit --ff $QUIET_ARGS "origin/$UPSTREAM_BRANCH"
+  fi
+
+  trap - SIGINT
+
+  CURRENT_REVISION="$(read_current_revision)"
+
+  if [ "$INITIAL_BRANCH" != "$UPSTREAM_BRANCH" ] && [ -n "$INITIAL_BRANCH" ]
+  then
+    git checkout "$INITIAL_BRANCH" $QUIET_ARGS
+    pop_stash
+  else
+    pop_stash_message
+  fi
+}
+
+update() {
+  for i in "$@"
+  do
+    case "$i" in
+      update) shift ;;
+      --verbose|-v) HOMEBREW_VERBOSE=1 ;;
+      --rebase)  HOMEBREW_REBASE=1 ;;
+      --simulate-from-current-branch) HOMEBREW_SIMULATE_FROM_CURRENT_BRANCH=1 ;;
+      --*) ;;
+      *)
+        echo "This command updates brew itself, and does not take formula names." >&2
+        echo "Use 'brew upgrade <formula>'." >&2
+        exit 1
+        ;;
+    esac
+  done
+
+  # check permissions
+  if [ "$HOMEBREW_PREFIX" = "/usr/local" ] && ! test -w /usr/local
+  then
+    echo "Error: /usr/local must writable!" >&2
+    exit 1
+  fi
+
+  if ! test -w "$HOMEBREW_REPOSITORY"
+  then
+    echo "Error: $HOMEBREW_REPOSITORY must writable!" >&2
+    exit 1
+  fi
+
+  if [ -z "$(which_git)" ]
+  then
+    brew install git
+    if [ -z "$(which_git)" ]
+    then
+      echo "Error: Git must be installed and in your PATH!" >&2
+    fi
+  fi
+
+  if [ -z "$HOMEBREW_VERBOSE" ]
+  then
+    QUIET_ARGS="-q"
+  fi
+
+  # ensure GIT_CONFIG is unset as we need to operate on .git/config
+  unset GIT_CONFIG
+
+  cd "$HOMEBREW_REPOSITORY" || {
+    echo "Error: failed to cd to $HOMEBREW_REPOSITORY!" >&2
+    exit 1
+  }
+  git_init_if_necessary
+
+  for DIR in "$HOMEBREW_REPOSITORY" $HOMEBREW_LIBRARY/Taps/*/*
+  do
+    [ -d "$DIR/.git" ] || continue
+    cd "$DIR" || continue
+    TAP_VAR=$(repo_var "$DIR")
+    export HOMEBREW_UPDATE_BEFORE"$TAP_VAR"="$(git rev-parse "$(upstream_branch)")"
+    UPSTREAM_BRANCH="$(upstream_branch)"
+    # the refspec ensures that the default upstream branch gets updated
+    git fetch $QUIET_ARGS origin \
+      "refs/heads/$UPSTREAM_BRANCH:refs/remotes/origin/$UPSTREAM_BRANCH" &
+  done
+
+  wait
+
+  for DIR in $HOMEBREW_REPOSITORY $HOMEBREW_LIBRARY/Taps/*/*
+  do
+    [ -d "$DIR/.git" ] || continue
+    pull "$DIR"
+  done
+
+  cd "$HOMEBREW_REPOSITORY" || {
+    echo "Error: failed to cd to $HOMEBREW_REPOSITORY!" >&2
+    exit 1
+  }
+}
+
+update "$@"
+exec "$HOMEBREW_RUBY_PATH" -W0 "$HOMEBREW_LIBRARY/brew.rb" report "$@"
```

Comment:
About `brew report $SHA1 $SHA1` idea, it would be useful. But `report` may be misleading considering we also do tap and formula rename migration. It may cause some surprise.